src/Admin/AjaxAction.php

```diff
@@ -0,0 +1,238 @@
+		callable $callback,
+		$selector = '',
+		$access = self::ACCESS_USER,
+		$scope = self::SCOPE_BACKEND
+	) {
+		if ( ! is_int( $access ) || $access < self::ACCESS_ANY || $access > self::ACCESS_NO_USER ) {
+			$access = self::ACCESS_USER;
+		}
+
+		if ( ! is_int( $scope ) || $scope < self::SCOPE_BOTH || $scope > self::SCOPE_FRONTEND ) {
+			$scope = self::SCOPE_BACKEND;
+		}
+
+		$this->action   = $action;
+		$this->callback = $callback;
+		$this->selector = $selector;
+		$this->access   = $access;
+		$this->scope    = $scope;
+	}
+
+	/**
+	 * Register the AJAX action with the WordPress system.
+	 */
+	public function register() {
+		if ( in_array( $this->access, [ self::ACCESS_USER, self::ACCESS_ANY ], true ) ) {
+			add_action( $this->get_action( self::ACCESS_USER ), $this->callback );
+		}
+
+		if ( in_array( $this->access, [ self::ACCESS_NO_USER, self::ACCESS_ANY ], true ) ) {
+			add_action( $this->get_action( self::ACCESS_NO_USER ), $this->callback );
+		}
+
+		if ( in_array( $this->scope, [ self::SCOPE_FRONTEND, self::SCOPE_BOTH ], true )
+			&& ! has_action( static::FRONTEND_ENQUEUE_ACTION, [ $this, 'register_ajax_script' ] ) ) {
+			add_action( static::FRONTEND_ENQUEUE_ACTION, [ $this, 'register_ajax_script' ] );
+		}
+
+		if ( in_array( $this->scope, [ self::SCOPE_BACKEND, self::SCOPE_BOTH ], true )
+			&& ! has_action( static::BACKEND_ENQUEUE_ACTION, [ $this, 'register_ajax_script' ] ) ) {
+			add_action( static::BACKEND_ENQUEUE_ACTION, [ $this, 'register_ajax_script' ] );
+		}
+	}
+
+	/**
+	 * Register the AJAX logic.
+	 */
+	public function register_ajax_script() {
+		if ( empty( $this->selector ) ) {
+			return;
+		}
+
+		if ( self::SCOPE_FRONTEND === $this->scope && is_admin() ) {
+			return;
+		}
+
+		if ( self::SCOPE_BACKEND === $this->scope && ! is_admin() ) {
+			return;
+		}
+
+		// phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
+		ob_start();
+
+		?>
+		<script>
+			( function ( $ ) {
+				$( function () {
+					var admin_ajax_url = '<?php echo addslashes( admin_url( 'admin-ajax.php' ) ); ?>';
+					var selector       = '<?php echo addslashes( $this->selector ); ?>';
+					setTimeout( function () {
+						( document.querySelectorAll( selector ) || [] )
+							.forEach( ( element ) => {
+								element.addEventListener( 'click', function ( event ) {
+									event.preventDefault();
+									$.ajax( {
+										type: "post",
+										dataType: "json",
+										url: admin_ajax_url,
+										data: { action: '<?php echo addslashes( $this->action ); ?>' }
+									}, { action: '<?php echo addslashes( $this->action ); ?>' } )
+										.done( function () {
+											element.classList.remove( 'ajax-failure' );
+											element.classList.add( 'ajax-success' )
+										} )
+										.fail( function () {
+											element.classList.remove( 'ajax-success' );
+											element.classList.add( 'ajax-failure' )
+										} );
+								} );
+							} );
+					}, 1000 );
+				} );
+			} )( jQuery );
```

Comment:
```suggestion
			} )();
```

Not needed if using `wp.ajax`.

Reply:
Resolved via https://github.com/ampproject/amp-wp/pull/4522/commits/d1e2aa21ddf73683daa9e92c66d87cfd0056c09d